### PR TITLE
Custom client support

### PR DESCRIPTION
**Pull request by BrennanConroy**
@sebastienros

### CODE REVIEW

**Comment by BrennanConroy** [comment]
@sebastienros could you take a look?
I'd like to iterate on it

File: src/Benchmarks.ClientJob/ClientJob.cs [hunk 1]
```diff
@@ -18,11 +18,11 @@ public ClientJob()
         public ClientJob(ClientJob clientJob)
         {
             Id = clientJob.Id;
-            Threads = clientJob.Threads;
             Connections = clientJob.Connections;
             Duration = clientJob.Duration;
-            ScriptName = clientJob.ScriptName;
-            PipelineDepth = clientJob.PipelineDepth;
+            MaxDuration = clientJob.MaxDuration;
```
Comment (sebastienros) [inline src/Benchmarks.ClientJob/ClientJob.cs]: We don't need it anymore

File: src/Benchmarks.ClientJob/ClientJob.cs [hunk 2]
```diff
@@ -35,16 +35,16 @@ public ClientJob(ClientJob clientJob)
...
-        public int Threads { get; set; } = 32;
+        public string ClientName { get; set; }
 
         public int Connections { get; set; } = 256;
 
         public int Duration { get; set; } = 15;
         public int Warmup { get; set; } = 15;
 
-        public string ScriptName { get; set; }
+        public int MaxDuration { get; set; } = 20;
```
Comment (sebastienros) [inline src/Benchmarks.ClientJob/ClientJob.cs]: Remove

File: src/BenchmarksClient/Startup.cs [hunk 3]
```diff
@@ -103,58 +100,42 @@ public static int Main(string[] args)
...
                         try
                         {
-                            Debug.Assert(process == null);
-
-                            await MeasureFirstRequestLatencyAsync(job);
+                            if (WorkerFactory.TryCreate(job, _httpClient, out worker, out var error) == false)
+                            {
+                                Console.WriteLine(error);
+                                // Worker failed to start
+                                job.State = ClientState.Completed;
```
Comment (sebastienros) [inline src/BenchmarksClient/Startup.cs]: `job.State = ClientState.Deleting;`

File: src/BenchmarksClient/Startup.cs [hunk 4]
```diff
@@ -103,58 +100,42 @@ public static int Main(string[] args)
...
-                            job.State = ClientState.Running;
-                            _jobs.Update(job);
+                                await worker.StartAsync();
 
-                            job.LastDriverCommunicationUtc = DateTime.UtcNow;
+                                Log($"Running job {worker.JobLogText}");
+                            }
 
-                            process = StartProcess(job);
+                            job.LastDriverCommunicationUtc = DateTime.Now;
```
Comment (sebastienros) [inline src/BenchmarksClient/Startup.cs]: `UtcNow`

File: src/BenchmarksClient/Startup.cs [hunk 5]
```diff
@@ -167,8 +148,7 @@ public static int Main(string[] args)
                     {
                         var now = DateTime.UtcNow;
 
-                        // Clean the job in case the driver is not running
-                        if (now - job.LastDriverCommunicationUtc > TimeSpan.FromSeconds(30))
+                        if (now - job.LastDriverCommunicationUtc > TimeSpan.FromSeconds(job.MaxDuration))
```
Comment (sebastienros) [inline src/BenchmarksClient/Startup.cs]: Revert to 30, this actually has nothing to do with `MaxDuration`. Later I'll update it to a much smaller value as the driver is actually communicating every second.

File: src/BenchmarksClient/Startup.cs [hunk 6]
```diff
@@ -177,21 +157,19 @@ public static int Main(string[] args)
...
 
                         try
                         {
-                            if (!process.HasExited)
-                            {
-                                process.Kill();
-                            }
+                            // Wait 5 seconds before killing worker
+                            var stopTask = worker.StopAsync();
+                            await Task.WhenAny(stopTask, Task.Delay((int)TimeSpan.FromSeconds(5).TotalMilliseconds));
```
Comment (sebastienros) [inline src/BenchmarksClient/Startup.cs]: We should really kill the worker. So `StopAsync` should itself be responsible for killing the process if it's not stopped after some delay. So no need to wait for anything else than `StopAsync`

File: src/BenchmarksClient/WorkerFactory.cs [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+using System.Net.Http;
+using Benchmarks.ClientJob;
+using BenchmarksClient.Workers;
+
+namespace BenchmarksClient
+{
+    static public class WorkerFactory
+    {
+        static public bool TryCreate(ClientJob clientJob, HttpClient httpClient, out IWorker worker, out string errors)
```
Comment (sebastienros) [inline src/BenchmarksClient/WorkerFactory.cs]: Do we really need the `HttpClient` ? This can be local to each worker.

File: src/BenchmarksServer/Startup.cs [hunk 8]
```diff
@@ -851,7 +851,9 @@ private static async Task<(string benchmarkDir, string dotnetDir)> CloneRestoreA
 
                 var outputFolder = Path.Combine(benchmarkedApp, "published");
 
-                ProcessUtil.Run(dotnetExecutable, $"publish -c Release -o {outputFolder} {buildParameters}",
+                var rid = OperatingSystem == OperatingSystem.Windows ? "win-x64" : "linux-x64";
```
Comment (sebastienros) [inline src/BenchmarksServer/Startup.cs]: Revert this change, I reverted it in the meantime, sorry.

File: src/BenchmarksDriver/Program.cs [hunk 9]
```diff
@@ -1193,7 +1211,7 @@ private static Task WriteJobsToSql(ServerJob serverJob, ClientJob clientJob, str
             string connectionFilter,
             WebHost webHost,
             int? kestrelThreadCount,
-            int clientThreads,
+            int? clientThreads,
```
Comment (sebastienros) [inline src/BenchmarksDriver/Program.cs]: Don't change that, just initialize to zero. We really want a value.

**Comment by BrennanConroy** [comment]
🆙 📅

File: src/BenchmarksDriver/Program.cs [hunk 10]
```diff
@@ -1325,7 +1337,7 @@ private static Task WriteJobsToSql(ServerJob serverJob, ClientJob clientJob, str
                         string.IsNullOrEmpty(connectionFilter) ? (object)DBNull.Value : connectionFilter);
                     p.AddWithValue("@WebHost", webHost.ToString());
                     p.AddWithValue("@KestrelThreadCount", (object)kestrelThreadCount ?? DBNull.Value);
-                    p.AddWithValue("@ClientThreads", clientThreads);
+                    p.AddWithValue("@ClientThreads", (object)clientThreads);
```
Comment (sebastienros) [inline src/BenchmarksDriver/Program.cs]: I assume that anything that is specific to workers can't be mapped to specific database columns.
But we should have a generic column that records the parameters in case we want to analyze it. Then we can just add an NVARCHAR(max) column that would contain a CSV with the worker properties. CSV would be easier to manipulate than Json in a db column.

File: src/BenchmarksDriver/Program.cs [hunk 11]
```diff
@@ -508,16 +519,17 @@ public static int Main(string[] args)
                 }
                 if (pipelineDepthOption.HasValue())
                 {
-                    _clientJob.PipelineDepth = int.Parse(pipelineDepthOption.Value());
+                    var pipelineDepth = int.Parse(pipelineDepthOption.Value());
+                    _clientJob.WorkerProperties["PipelineDepth"] = pipelineDepth;
```
Comment (sebastienros) [inline src/BenchmarksDriver/Program.cs]: This could be handled in a generic way were we pass client specific arguments to the worker properties.
Something like `--client wrk --properties thread=16;connections=32;pipelinedepth=16` or similar

Comment (BrennanConroy) [reply src/BenchmarksDriver/Program.cs]: Do we want to get rid of the `--clientThreads` `--script` and `--pipelineDepth` options

**Comment by BrennanConroy** [comment]
Rebased and :up: 📅

**Comment by davidfowl** [comment]
@sebastienros can you review this please? We're trying to get performance runs for SignalR going ASAP.

**Comment by davidfowl** [comment]
:clap:
